### PR TITLE
Add & args to methods to allow passing `keys` argument

### PR DESCRIPTION
**Pull request by WhittlesJr**
My attempt at solving [this issue](https://github.com/weavejester/integrant-repl/issues/9).

Is `& args` acceptable? I'm not a fan it (the ambiguity, the verbosity of the apply/partial stuff), but... I started doing explicit overloading like Integrant does, but it got weird in places... This was much easier to write for me.

### CODE REVIEW

**Comment by WhittlesJr** [comment]
Oh, I should mention that I was at a loss with the reset and reset-all functions, so I left them as-is. :after has to be a symbol, and the invocation of it  in clojure.tools.namespace.repl/do-refresh doesn't give you any ability to pass other arguments.

**Comment by weavejester** [comment]
You shouldn't use `& args` if you have a finite argument list. Use explicit arities instead. Explicit arities are both more efficient, and will raise errors when the argument number is incorrect.

`reset` and `reset-all` can be passed keys through the `integrant.repl.state` namespace.

**Comment by WhittlesJr** [comment]
I'll switch back to explicit arities then. 

The thought had crossed my mind to set keys for reset and reset-all on an atom or something as you suggest... but it cringed me out too much. I don't see a better way in this case though...

**Comment by WhittlesJr** [comment]
Is that better? Like I said, it's a bit awkward, which is why I avoided it at first. But I agree that this is preferable on principle.

File: src/integrant/repl.clj [hunk 1]
```diff
@@ -16,8 +16,16 @@
     (do (alter-var-root #'state/config (fn [_] (prep))) :prepped)
     (throw (prep-error))))
 
-(defn- halt-system [system]
-  (when system (ig/halt! system)))
+(defn all-system-keys
```
Comment (weavejester) [inline src/integrant/repl.clj]: I don't think this function is necessary if `nil` is used to denote every key should be used.

File: src/integrant/repl.clj [hunk 2]
```diff
@@ -16,8 +16,16 @@
...
+  []
+  (keys state/config))
+
+(defn- halt-system
+  ([system]
+   (halt-system (keys system)))
+
```
Comment (weavejester) [inline src/integrant/repl.clj]: Blank lines shouldn't be placed between arities definitions.

File: src/integrant/repl.clj [hunk 3]
```diff
@@ -16,8 +16,16 @@
     (do (alter-var-root #'state/config (fn [_] (prep))) :prepped)
     (throw (prep-error))))
 
-(defn- halt-system [system]
-  (when system (ig/halt! system)))
+(defn all-system-keys
+  []
+  (keys state/config))
+
+(defn- halt-system
```
Comment (weavejester) [inline src/integrant/repl.clj]: This is a private function; it doesn't need two arities.

File: src/integrant/repl.clj [hunk 4]
```diff
@@ -29,59 +37,110 @@
...
-    (throw (prep-error))))
-
-(defn reset []
-  (suspend)
-  (repl/refresh :after 'integrant.repl/resume))
-
-(defn reset-all []
-  (suspend)
-  (repl/refresh-all :after 'integrant.repl/resume))
+(defn- init-system
```
Comment (weavejester) [inline src/integrant/repl.clj]: Private function doesn't need two arities.

File: src/integrant/repl.clj [hunk 5]
```diff
@@ -29,59 +37,110 @@
...
+   (init-system config (keys config)))
+
+  ([config keys]
+   (build-system
+    #(ig/init config keys)
+    #(ex-info "Config failed to init; also failed to halt failed system"
+              {:init-exception %1}
+              %2))))
+
+(defn- resume-system
```
Comment (weavejester) [inline src/integrant/repl.clj]: Ditto about private functions and unncessary whitespace.

File: src/integrant/repl.clj [hunk 6]
```diff
@@ -29,59 +37,110 @@
...
+   (resume-system config system (keys config)))
+
+  ([config system keys]
+   (build-system
+    #(ig/resume config system keys)
+    #(ex-info "Config failed to resume; also failed to halt failed system"
+              {:resume-exception %1}
+              %2))))
+
+(defn init
```
Comment (weavejester) [inline src/integrant/repl.clj:49]: Something more like:

```clojure
(defn init
  ([] (init nil))
  ([keys]
   (alter-var-root #'state/system (fn [sys]
                                    (halt-system sys keys)
                                    (init-system state/config keys)))
```

Then `halt-system` and `init-system` can check for the presence of `nil`.

File: src/integrant/repl.clj [hunk 7]
```diff
@@ -29,59 +37,110 @@
...
+  ([]
+   (init (all-system-keys)))
+
+  ([keys]
+   (alter-var-root #'state/system (fn [sys]
+                                    (halt-system sys keys)
+                                    (init-system state/config keys)))
+   :initiated))
+
+(defn go
```
Comment (weavejester) [inline src/integrant/repl.clj:57]: Something more like:

```clojure
(defn go
  ([] (go nil))
  ([keys]
   (prep)
   (init keys)))
```

File: src/integrant/repl/state.clj [hunk 8]
```diff
@@ -6,3 +6,4 @@
 (def config   nil)
 (def system   nil)
 (def preparer nil)
+(def selected-keys-for-reset nil)
```
Comment (weavejester) [inline src/integrant/repl/state.clj]: Change this to `initiated-keys`

File: src/integrant/repl.clj [hunk 9]
```diff
@@ -29,59 +37,110 @@
...
+           keys (or state/selected-keys-for-reset keys)]
+       (alter-var-root #'state/config (constantly cfg))
+       (alter-var-root #'state/system (fn [sys]
+                                        (if sys
+                                          (resume-system cfg sys keys)
+                                          (init-system cfg keys))))
+       :resumed)
+     (throw (prep-error)))))
+
+(defn- set-selected-keys!
```
Comment (weavejester) [inline src/integrant/repl.clj]: Reduce the whitespace:

```clojure
(defn- set-initiated-keys! [keys]
  (alter-var-root #'state/initiated-keys (constantly keys)))
```

**Review by weavejester — CHANGES_REQUESTED** [review]

File: src/integrant/repl.clj [hunk 10]
```diff
@@ -29,59 +32,83 @@
...
+       :resumed)
+     (throw (prep-error)))))
+
+(defn- set-selected-keys! [keys]
+  (alter-var-root #'state/initiated-keys (constantly keys)))
+
+(defn reset
+  ([] (reset nil))
+  ([keys]
+   (set-selected-keys! keys)
```
Comment (weavejester) [inline src/integrant/repl.clj]: Should this be in `reset`, or should we have something like `set-suspended-keys`?

For that matter, do we need the var at all? Surely `(keys system)` would amount to the same thing?

**Review by weavejester — CHANGES_REQUESTED** [review]

**Comment by weavejester** [comment]
Thanks for the changes. I think we might be able to remove `initiated-keys` and replace it with `(keys system)`, unless I'm not thinking of something.

**Comment by WhittlesJr** [comment]
I don't believe that will work... How would you pass keys from `reset`? It would always reset the entire system, wouldn't it? Because `resume` would be called from `repl/refresh` with no args, so `keys` would be `nil`, and `resume-system` would infer that it needs to operate on all keys of the system. 

Unless you see something I don't...

**Comment by weavejester** [comment]
Ah, I was only considering the case where keys were on `init` or possibly `halt`. However, that does suggest the question of whether we want keys on all the functions, or just on some of them. I can see a use-case for keys on `init` and `go`, but what's the use-case for the other functions?

**Comment by WhittlesJr** [comment]
Well, you have `keys` on those functions in Integrant, so I figured you had use cases for them already. Unless that was never intended for REPL use?

**Comment by weavejester** [comment]
Integrant has is a low-level framework, and it's difficult to predict how people will use it. Integrant-REPL on the other hand is a convenience library that is designed to guide the user with a specific workflow. I assumed you were implementing this to solve a problem you were having, but I guess I should have asked first! :)

**Comment by WhittlesJr** [comment]
I guess for my case, what I needed was to init one key (+ dependencies) for debugging purposes, to try to isolate the problem I was having. So what I care about is init, halt, and reset for specified keys. I don't personally use resume or suspend, but adding this feature for those three methods helped me tremendously.

**Comment by weavejester** [comment]
Were the keys you used for `halt` and `reset` different to the keys you used for `init`?

**Comment by WhittlesJr** [comment]
Sometimes kind of... The case I can think of is this: I init the system subset, I poke around, change something, then reset just that subset to see how it changed. It changed favorably, so now I want to reset again and get the full system back, so I call reset with no args. So in that case, not tying reset to only what I init'ed is helpful. 

Granted, we could do what you suggest (having reset operate only on what was init'ed), and say that the user should do a halt and init with no args to get the full system back.

Totally up to you.

**Comment by weavejester** [comment]
Thanks for detailing your use-case. I was approaching it from the idea that you could `init` a subset, but the `reset` etc. would follow that subset of keys through.

I'll sleep on this, but I'm leaning toward allowing this PR to go through. My only concern is that it doesn't enforce a consistent state: we could potentially have a system with some keys initiated, some keys halted, and some keys suspended.

**Comment by WhittlesJr** [comment]
The more I think about it, the more I like enforcing a consistent state and workflow. Like you said, that's pretty much what this library is all about. I'll be happy either way.

**Comment by WhittlesJr** [comment]
So I've noticed that with these changes, `reset` doesn't first halt the system... looking into that

**Comment by weavejester** [comment]
I still haven't entirely decided on whether this is a good idea or not (on the one hand it adds some useful options, but on the other it complicates the codebase). Some other changes to Integrant and Duct have taken my attention, so this PR might sit on the back burner for a while.

**Comment by WhittlesJr** [comment]
No worries. I just pushed changes that make suspend / resume / reset / reset-all / halt use only the keys that have been init'ed. So if you init with a subset, it resets/suspends/etc. with the same subset. 

This also fixed the aforementioned issue with reset not halting things.

**Comment by WhittlesJr** [comment]
And just as one final plea, I'll add that this ability has been indispensable for me in debugging a large and complex system. It's often critical for me to init only keys that will come up without throwing exceptions so I can see what the system looks like, and what might be causing the errors I see in the rest of the keys.

**Comment by weavejester** [comment]
I've looked over the changes you've added. I'm much more amenable to just having keys on `init` (and `go`), since that eliminates any chance of inconsistency.

I believe it also does away with the `initiated-keys` var, since a system contains only the initiated keys. I think you could remove things like `initiated-keys` and `get-keys` from your code and simplify it further, since `(= initiated-keys (keys system))`.

**Comment by madstap** [comment]
This came up at work today and lead me to this PR. It seems like it stalled. I'd be happy to finish implementing this, if it's still wanted?

**Comment by WhittlesJr** [comment]
I'm not sure what is still needed. I'd love to get this mergeable.

**Comment by weavejester** [comment]
My previous comment details what's needed to get this PR merged; all that's left is implementation. If anything isn't clear, feel free to ask.

**Comment by WhittlesJr** [comment]
Oh, sorry, I thought I had done that stuff. How does this look?

File: src/integrant/repl.clj [hunk 11]
```diff
@@ -44,18 +46,22 @@
...
-  (prep)
-  (init))
+(defn go
+  ([] (go nil))
+  ([keys]
+   (prep)
+   (init keys)))
 
 (defn clear []
-  (alter-var-root #'state/system (fn [sys] (halt-system sys) nil))
```
Comment (madstap) [inline src/integrant/repl.clj:58]: This looks like a typo

Comment (weavejester) [reply src/integrant/repl.clj:58]: Yes, I think this change is probably a mistake.

File: src/integrant/repl.clj [hunk 12]
```diff
@@ -30,9 +30,11 @@
             (throw (wrap-ex ex halt-ex)))))
       (throw ex))))
 
-(defn- init-system [config]
+(defn- init-system [config & [keys]]
```
Comment (weavejester) [inline src/integrant/repl.clj]: As this is a private function, I don't think `keys` needs to be an optional argument.

File: src/integrant/repl.clj (hunk 11 above)
Comment (WhittlesJr) [reply src/integrant/repl.clj:58]: Yup, no idea how that happened. I'll fix it soon.

Comment (WhittlesJr) [reply src/integrant/repl.clj:58]: Weird, my local copy doesn't even show this line changing or being deleted, and there's no changes visible in git... I'm really not sure what happened or even how to fix it. Maybe I'll try squashing

Comment (WhittlesJr) [reply src/integrant/repl.clj:58]: Even after rebasing on master and squashing everything into one commit, this line still appears deleted on GitHub, but not on my local machine.

Comment (WhittlesJr) [reply src/integrant/repl.clj:58]: Could you pull my branch and confirm?

Comment (WhittlesJr) [reply src/integrant/repl.clj:58]: Er, wait, sorry, I'm just blind. I thought the line was deleted but it was just a paren that was moved.

**Comment by madstap** [comment]
It looks like all the requested changes were implemented, is there anything still missing?

**Comment by weavejester** [comment]
Thanks for the update. As a quick note, Github doesn't notify me when a commit is updated, only when a comment is made, so I won't know if you've made any changes unless you poke me with a comment :)

The PR looks good. Can you change the commit message to:

     Add 'keys' argument to init and go functions

**Comment by psdp** [comment]
Hi, any update on this? This is a very useful feature indeed.

**Comment by weavejester** [comment]
The PR is waiting on changes from the author.

**Comment by WhittlesJr** [comment]
Commit message updated. Sorry for the wait.

**Pull request merged**